File: rental/src/main/java/io/axoniq/demo/bikerental/rental/ui/BikeRentalDataGenerator.java
package io.axoniq.demo.bikerental.rental.ui;

import io.axoniq.demo.bikerental.coreapi.payment.ConfirmPaymentCommand;
import io.axoniq.demo.bikerental.coreapi.payment.PaymentStatusNamedQueries;
import io.axoniq.demo.bikerental.coreapi.rental.BikeStatus;
import io.axoniq.demo.bikerental.coreapi.rental.BikeStatusNamedQueries;
import io.axoniq.demo.bikerental.coreapi.rental.RegisterBikeCommand;
import io.axoniq.demo.bikerental.coreapi.rental.RentalStatus;
import io.axoniq.demo.bikerental.coreapi.rental.RequestBikeCommand;
import io.axoniq.demo.bikerental.coreapi.rental.ReturnBikeCommand;
import io.axoniq.demo.bikerental.rental.support.CommandDispatcher;
import io.axoniq.demo.bikerental.rental.support.QueryDispatcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

@Component
public class BikeRentalDataGenerator {

    private static final List<String> RENTERS = Arrays.asList("Allard", "Steven", "Josh", "David", "Marc", "Sara", "Milan", "Jeroen", "Marina", "Jeannot");
    private static final List<String> LOCATIONS = Arrays.asList("Amsterdam", "Paris", "Vilnius", "Barcelona", "London", "New York", "Toronto", "Berlin", "Milan", "Rome", "Belgrade");

    private final CommandDispatcher commandDispatcher;
    private final QueryDispatcher queryDispatcher;

    Logger logger = LoggerFactory.getLogger(BikeRentalDataGenerator.class);

    public BikeRentalDataGenerator(CommandDispatcher commandDispatcher, QueryDispatcher queryDispatcher) {
        this.commandDispatcher = commandDispatcher;
        this.queryDispatcher = queryDispatcher;
    }

    public String randomRenter() {
        return RENTERS.get(ThreadLocalRandom.current().nextInt(RENTERS.size()));
    }

    public String randomLocation() {
        return LOCATIONS.get(ThreadLocalRandom.current().nextInt(LOCATIONS.size()));
    }

    public CompletableFuture<Void> generateBikes(int bikeCount, String bikeType) {
        CompletableFuture<Void> all = CompletableFuture.completedFuture(null);
        for (int i = 0; i < bikeCount; i++) {
            all = CompletableFuture.allOf(all,
                    commandDispatcher.send(new RegisterBikeCommand(UUID.randomUUID().toString(), bikeType, randomLocation())));
        }
        return all;
    }

    public Flux<String> generateRentals(String bikeType, int loops, int concurrency, int abandonPaymentFactor, int delay) {
        return Flux.range(0, loops)
                   .flatMap(j -> executeRentalCycle(bikeType, randomRenter(), abandonPaymentFactor, delay)
                                   .map(r -> "OK - Rented, Payed and Returned\n")
                                   .onErrorResume(e -> Mono.just("Not ok: " + e.getMessage() + "\n")),
                            concurrency);
    }

    private Mono<String> executeRentalCycle(String bikeType, String renter, int abandonPaymentFactor, int delay) {
        CompletableFuture<String> result = selectRandomAvailableBike(bikeType)
                .thenCompose(bikeId -> commandDispatcher.send(new RequestBikeCommand(bikeId, renter), String.class)
                        .thenComposeAsync(paymentRef -> executePayment(bikeId, paymentRef, abandonPaymentFactor),
                                CompletableFuture.delayedExecutor(randomDelay(delay), TimeUnit.MILLISECONDS))
                        .thenCompose(r -> whenBikeUnlocked(bikeId))
                        .thenComposeAsync(r -> commandDispatcher.send(new ReturnBikeCommand(bikeId, randomLocation())),
                                CompletableFuture.delayedExecutor(randomDelay(delay), TimeUnit.MILLISECONDS))
                        .thenApply(r -> bikeId));
        return Mono.fromFuture(result);
    }

    private CompletableFuture<String> selectRandomAvailableBike(String bikeType) {
        return queryDispatcher.queryMany(BikeStatusNamedQueries.FIND_AVAILABLE, bikeType, BikeStatus.class)
                              .thenApply(this::pickRandom)
                              .thenApply(BikeStatus::getBikeId);
    }

    private CompletableFuture<String> executePayment(String bikeId, String paymentRef, int abandonPaymentFactor) {
        if (abandonPaymentFactor > 0 && ThreadLocalRandom.current().nextInt(abandonPaymentFactor) == 0) {
            return CompletableFuture.failedFuture(new IllegalStateException("Customer refused to pay"));
        }
        return queryDispatcher.subscriptionQuery(PaymentStatusNamedQueries.GET_PAYMENT_ID, paymentRef, String.class)
                              .filter(Objects::nonNull)
                              .next()
                              .flatMap(paymentId -> Mono.fromFuture(commandDispatcher.send(new ConfirmPaymentCommand(paymentId))))
                              .map(o -> bikeId)
                              .toFuture();
    }

    private CompletableFuture<String> whenBikeUnlocked(String bikeId) {
        return queryDispatcher.subscriptionQuery(BikeStatusNamedQueries.FIND_ONE, bikeId, BikeStatus.class)
                              .any(status -> status.getStatus() == RentalStatus.RENTED)
                              .map(s -> bikeId)
                              .toFuture();
    }

    private int randomDelay(int delay) {
        if (delay <= 0) {
            return 0;
        }
        return ThreadLocalRandom.current().nextInt(delay - (delay >> 2), delay + delay + (delay >> 2));
    }

    private <T> T pickRandom(List<T> source) {
        return source.get(ThreadLocalRandom.current().nextInt(source.size()));
    }
}

File: rental/src/ui/bike-rental-data-generator.ts
import { randomUUID } from 'node:crypto';
import {
  BikeStatus,
  BikeStatusNamedQueries,
  CommandDispatcher,
  ConfirmPaymentCommand,
  JavaTypes,
  PaymentStatusNamedQueries,
  QueryDispatcher,
  RegisterBikeCommand,
  RentalStatus,
  RequestBikeCommand,
  ReturnBikeCommand,
} from '@bikerental/core-api';

const RENTERS = ['Allard', 'Steven', 'Josh', 'David', 'Marc', 'Sara', 'Milan', 'Jeroen', 'Marina', 'Jeannot'];
const LOCATIONS = ['Amsterdam', 'Paris', 'Vilnius', 'Barcelona', 'London', 'New York', 'Toronto', 'Berlin', 'Milan', 'Rome', 'Belgrade'];

const delay = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

function randomInt(boundExclusive: number): number {
  return Math.floor(Math.random() * boundExclusive);
}

function randomIntBetween(originInclusive: number, boundExclusive: number): number {
  return originInclusive + Math.floor(Math.random() * (boundExclusive - originInclusive));
}

export class BikeRentalDataGenerator {
  constructor(
    private readonly commandDispatcher: CommandDispatcher,
    private readonly queryDispatcher: QueryDispatcher,
  ) {}

  randomRenter(): string {
    return RENTERS[randomInt(RENTERS.length)];
  }

  randomLocation(): string {
    return LOCATIONS[randomInt(LOCATIONS.length)];
  }

  async generateBikes(bikeCount: number, bikeType: string): Promise<void> {
    const sends: Promise<void>[] = [];
    for (let i = 0; i < bikeCount; i++) {
      sends.push(
        this.commandDispatcher.send(new RegisterBikeCommand(randomUUID(), bikeType, this.randomLocation())),
      );
    }
    await Promise.all(sends);
  }

  /**
   * Runs `loops` rental cycles with at most `concurrency` in flight, reporting one result line per
   * cycle through `onLine` (mirroring the Flux the Java controller streamed back).
   */
  async generateRentals(
    bikeType: string,
    loops: number,
    concurrency: number,
    abandonPaymentFactor: number,
    delayMs: number,
    onLine: (line: string) => void,
  ): Promise<void> {
    let next = 0;
    const worker = async (): Promise<void> => {
      for (;;) {
        const index = next++;
        if (index >= loops) {
          return;
        }
        try {
          await this.executeRentalCycle(bikeType, this.randomRenter(), abandonPaymentFactor, delayMs);
          onLine('OK - Rented, Payed and Returned\n');
        } catch (error) {
          onLine(`Not ok: ${(error as Error).message}\n`);
        }
      }
    };
    const workers = Array.from({ length: Math.max(1, Math.min(concurrency, loops)) }, worker);
    await Promise.all(workers);
  }

  private async executeRentalCycle(
    bikeType: string,
    renter: string,
    abandonPaymentFactor: number,
    delayMs: number,
  ): Promise<string> {
    const bikeId = await this.selectRandomAvailableBike(bikeType);
    const paymentRef = (await this.commandDispatcher.send<string>(
      new RequestBikeCommand(bikeId, renter),
      JavaTypes.STRING,
    ))!;
    await delay(this.randomDelay(delayMs));
    await this.executePayment(bikeId, paymentRef, abandonPaymentFactor);
    await this.whenBikeUnlocked(bikeId);
    await delay(this.randomDelay(delayMs));
    await this.commandDispatcher.send(new ReturnBikeCommand(bikeId, this.randomLocation()));
    return bikeId;
  }

  private async selectRandomAvailableBike(bikeType: string): Promise<string> {
    const available = await this.queryDispatcher.queryMany<BikeStatus>(
      BikeStatusNamedQueries.FIND_AVAILABLE,
      bikeType,
      BikeStatus.TYPE,
    );
    return this.pickRandom(available).getBikeId();
  }

  private async executePayment(
    bikeId: string,
    paymentRef: string,
    abandonPaymentFactor: number,
  ): Promise<string> {
    if (abandonPaymentFactor > 0 && randomInt(abandonPaymentFactor) === 0) {
      throw new Error('Customer refused to pay');
    }
    const paymentId = await this.queryDispatcher.pollFor<string>(
      PaymentStatusNamedQueries.GET_PAYMENT_ID,
      paymentRef,
      JavaTypes.STRING,
      (result) => result !== null,
    );
    await this.commandDispatcher.send(new ConfirmPaymentCommand(paymentId));
    return bikeId;
  }

  private async whenBikeUnlocked(bikeId: string): Promise<string> {
    await this.queryDispatcher.pollFor<BikeStatus>(
      BikeStatusNamedQueries.FIND_ONE,
      bikeId,
      BikeStatus.TYPE,
      (status) => status?.getStatus() === RentalStatus.RENTED,
    );
    return bikeId;
  }

  private randomDelay(delayMs: number): number {
    if (delayMs <= 0) {
      return 0;
    }
    return randomIntBetween(delayMs - (delayMs >> 2), delayMs + delayMs + (delayMs >> 2));
  }

  private pickRandom<T>(source: T[]): T {
    if (source.length === 0) {
      throw new Error('No available bike to pick');
    }
    return source[randomInt(source.length)];
  }
}
